data/tutorials/wf_03_metaprogramming.md

```diff
@@ -0,0 +1,544 @@
+```
+
+Extension nodes are meant to be rewritten by a PPX and, in this regard,
+correspond to a specific kind of PPX: extenders. An extender is a PPX rewriter
+which will replace all occurrences of extension nodes with a matching name. It does this with 
+some generated code that depends only on the payload, without information on the
+context of the extension node (that is, without access to the rest of the code)
+and without modifying anything else.
+
+Examples of extenders include:
+- Extenders which allow users to write OCaml values representing another language 
+  directly in such language. For example:
+      - `ppx_yojson` to generate `Yojson` values by writing JSON code
+      - `tyxml-ppx` to generate `Tyxml.Html` values by writing HTML code
+      - `ppx_mysql` to generate `ocaml-mysql` queries by writing MySQL
+        queries
+- `ppx_expect` to generate CRAM tests from the extension node's payload.
+
+### Using PPXs
+
+Similarly to preprocessors for the text source code, OCaml's compiler provides
+the `-ppx` option to run a PPX during the compilation phase. The PPX will read
+the Parsetree from a file, where the marshalled value has been dumped, and output
+the rewritten Parsetree the same way.
+
+But again, since the tool responsible for driving the compilation is Dune, using
+PPXs is just a matter of writing `dune` files. The same `preprocess` stanza
+should be used, this time with `pps`:
+
+```dune
+  (preprocess (pps ppx1 ppx2))
+```
+
+And that's all you need to use PPXs! Although these instructions will work for
+most PPXs, note that the first source of information will be the package
+documentation, as some PPXs might need some special care to integrate with
+Dune.
+
+### Dropping PPXs dependency with `[@@deriving_inline]`
+
+Some derivers are only needed for boilerplate generation. When that's the case, 
+there is no strong requirement
+for them to be included as a hard dependency: the added boilerplate code can be pretty printed and added
+ to the source code by the PPX. This mechanism can be implemented using Dune
+and [`ppxlib`](#the-need-for-controlling-the-ppx-ecosystem-ppxlib).
+
+Attaching `[@@deriving_inline <deriver_name>]` to an item will derive some code
+from the item just like the usual `[@@deriving <deriver_name>]` attribute.
+However, instead of appending the generated code after the attributed item, it
+will check that the generated code is already present after the attributed item.
+If yes, nothing has to be done. Otherwise, it will generate a correct file, and
+Dune will offer you the possibility of updating your source code using this
+correct file (using the `dune promote` command).
+
+As the new file contains the generated code, it no longer needs to be
+preprocessed by the PPX, and can be compiled and distributed as is, and the PPX can be
+removed from the dependencies. However, the PPX still needs to be run whenever the
+item on which the attribute is attached change. This can be achieved by running
```

Comment:
```suggestion
removed from the dependencies. However, the PPX still needs to be run whenever the
item on which the attribute is attached changes. This can be achieved by running
```